File: Server/MultiplayerDemo.Server/Protocol.cs
using System.Text.Json.Serialization;

namespace MultiplayerDemo.Server;

// Wire shapes for docs/specs/26_09_08_09_multiplayer_arena/protocol.md.
// Field names are deliberately short: snapshots go out 30 times a second.

public sealed class ClientMessage {
	[JsonPropertyName("t")] public string T { get; set; } = string.Empty;
	[JsonPropertyName("name")] public string? Name { get; set; }
	[JsonPropertyName("client")] public string? Client { get; set; }
	[JsonPropertyName("mx")] public float Mx { get; set; }
	[JsonPropertyName("my")] public float My { get; set; }
	[JsonPropertyName("fire")] public bool Fire { get; set; }
	[JsonPropertyName("id")] public int Id { get; set; }
}

public sealed class ArenaBounds {
	[JsonPropertyName("minX")] public float MinX { get; set; }
	[JsonPropertyName("minY")] public float MinY { get; set; }
	[JsonPropertyName("maxX")] public float MaxX { get; set; }
	[JsonPropertyName("maxY")] public float MaxY { get; set; }
}

public sealed class WelcomeMessage {
	[JsonPropertyName("t")] public string T => "welcome";
	[JsonPropertyName("id")] public int Id { get; set; }
	[JsonPropertyName("tickRate")] public int TickRate { get; set; }
	[JsonPropertyName("arena")] public ArenaBounds Arena { get; set; } = new();
	[JsonPropertyName("playerRadius")] public float PlayerRadius { get; set; }
	[JsonPropertyName("bulletRadius")] public float BulletRadius { get; set; }
	[JsonPropertyName("moveSpeed")] public float MoveSpeed { get; set; }
	[JsonPropertyName("maxHealth")] public int MaxHealth { get; set; }
	[JsonPropertyName("respawnMs")] public int RespawnMs { get; set; }
	[JsonPropertyName("fireCooldownMs")] public int FireCooldownMs { get; set; }
	[JsonPropertyName("startedAtMs")] public long StartedAtMs { get; set; }
	[JsonPropertyName("serverNowMs")] public long ServerNowMs { get; set; }
}

public sealed class RejectMessage {
	[JsonPropertyName("t")] public string T => "reject";
	[JsonPropertyName("reason")] public string Reason { get; set; } = string.Empty;
}

public sealed class StatePlayer {
	[JsonPropertyName("id")] public int Id { get; set; }
	[JsonPropertyName("x")] public float X { get; set; }
	[JsonPropertyName("y")] public float Y { get; set; }
	[JsonPropertyName("hp")] public int Hp { get; set; }
	[JsonPropertyName("dx")] public float Dx { get; set; }
	[JsonPropertyName("dy")] public float Dy { get; set; }
	[JsonPropertyName("cd")] public int Cd { get; set; }
}

public sealed class StateBullet {
	[JsonPropertyName("id")] public int Id { get; set; }
	[JsonPropertyName("x")] public float X { get; set; }
	[JsonPropertyName("y")] public float Y { get; set; }
	[JsonPropertyName("o")] public int O { get; set; }
}

public sealed class StateMessage {
	[JsonPropertyName("t")] public string T => "state";
	[JsonPropertyName("tick")] public int Tick { get; set; }
	[JsonPropertyName("elapsedMs")] public long ElapsedMs { get; set; }
	[JsonPropertyName("players")] public List<StatePlayer> Players { get; set; } = new();
	[JsonPropertyName("bullets")] public List<StateBullet> Bullets { get; set; } = new();
}

public sealed class RosterPlayer {
	[JsonPropertyName("id")] public int Id { get; set; }
	[JsonPropertyName("name")] public string Name { get; set; } = string.Empty;
	[JsonPropertyName("kills")] public int Kills { get; set; }
	[JsonPropertyName("deaths")] public int Deaths { get; set; }
	[JsonPropertyName("connected")] public bool Connected { get; set; }
	[JsonPropertyName("ip")] public string Ip { get; set; } = string.Empty;
	[JsonPropertyName("client")] public string Client { get; set; } = string.Empty;
	[JsonPropertyName("ping")] public int Ping { get; set; }
	[JsonPropertyName("respawnIn")] public int RespawnIn { get; set; }
}

public sealed class RosterMessage {
	[JsonPropertyName("t")] public string T => "roster";
	[JsonPropertyName("players")] public List<RosterPlayer> Players { get; set; } = new();
}

public sealed class PingMessage {
	[JsonPropertyName("t")] public string T => "ping";
	[JsonPropertyName("id")] public int Id { get; set; }
}

public sealed class EventMessage {
	[JsonPropertyName("t")] public string T => "event";
	[JsonPropertyName("kind")] public string Kind { get; set; } = string.Empty;
	[JsonPropertyName("victim")] public int Victim { get; set; }
	[JsonPropertyName("killer")] public int Killer { get; set; }
}

public sealed class ServerInfo {
	[JsonPropertyName("name")] public string Name { get; set; } = string.Empty;
	[JsonPropertyName("players")] public int Players { get; set; }
	[JsonPropertyName("maxPlayers")] public int MaxPlayers { get; set; }
	[JsonPropertyName("version")] public int Version => 1;
	[JsonPropertyName("tickRate")] public int TickRate { get; set; }
	[JsonPropertyName("wsPath")] public string WsPath => "/ws";
	[JsonPropertyName("uptimeMs")] public long UptimeMs { get; set; }
}

public sealed class BeaconMessage {
	[JsonPropertyName("magic")] public string Magic => "MPDEMO1";
	[JsonPropertyName("name")] public string Name { get; set; } = string.Empty;
	[JsonPropertyName("port")] public int Port { get; set; }
	[JsonPropertyName("players")] public int Players { get; set; }
}

File: WebClient/src/protocol.ts
/**
 * Wire shapes from docs/specs/26_09_08_09_multiplayer_arena/protocol.md (v1).
 * Frozen contract — do not extend without updating that document first.
 */

export type ClientKind = 'unity' | 'web' | 'unknown';

export interface ArenaBounds {
	minX: number;
	minY: number;
	maxX: number;
	maxY: number;
}

export interface WelcomeMessage {
	t: 'welcome';
	id: number;
	tickRate: number;
	arena: ArenaBounds;
	playerRadius: number;
	bulletRadius: number;
	moveSpeed: number;
	maxHealth: number;
	respawnMs: number;
	fireCooldownMs: number;
	startedAtMs: number;
	serverNowMs: number;
}

export interface RejectMessage {
	t: 'reject';
	reason: string;
}

export interface StatePlayer {
	id: number;
	x: number;
	y: number;
	hp: number;
	dx: number;
	dy: number;
	cd: number;
}

export interface StateBullet {
	id: number;
	x: number;
	y: number;
	o: number;
}

export interface StateMessage {
	t: 'state';
	tick: number;
	elapsedMs: number;
	players: StatePlayer[];
	bullets: StateBullet[];
}

export interface RosterPlayer {
	id: number;
	name: string;
	kills: number;
	deaths: number;
	connected: boolean;
	ip: string;
	client: string;
	ping: number;
	respawnIn: number;
}

export interface RosterMessage {
	t: 'roster';
	players: RosterPlayer[];
}

export interface PingMessage {
	t: 'ping';
	id: number;
}

export interface EventMessage {
	t: 'event';
	kind: 'kill' | 'hit' | 'spawn';
	victim: number;
	killer: number;
}

export type ServerMessage =
	| WelcomeMessage
	| RejectMessage
	| StateMessage
	| RosterMessage
	| PingMessage
	| EventMessage;

export interface JoinMessage {
	t: 'join';
	name: string;
	client: 'unity' | 'web';
}

export interface InputMessage {
	t: 'input';
	mx: number;
	my: number;
	fire: boolean;
}

export interface PongMessage {
	t: 'pong';
	id: number;
}

export type ClientMessage = JoinMessage | InputMessage | PongMessage;

export interface ServerInfo {
	name: string;
	players: number;
	maxPlayers: number;
	version: number;
	tickRate: number;
	wsPath: string;
	uptimeMs: number;
}

/** Returns null instead of throwing: a malformed frame must never kill the render loop. */
export function parseServerMessage(raw: string): ServerMessage | null {
	let value: unknown;
	try {
		value = JSON.parse(raw);
	} catch {
		return null;
	}
	if ((typeof value !== 'object') || (value === null)) {
		return null;
	}
	const t = (value as { t?: unknown }).t;
	switch (t) {
		case 'welcome':
		case 'reject':
		case 'state':
		case 'roster':
		case 'ping':
		case 'event':
			return value as ServerMessage;
		default:
			return null;
	}
}
